nixos/modules/services/misc/nix-optimise.nix

```diff
@@ -0,0 +1,49 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.nix.optimise;
+in
+
+{
+
+  ###### interface
+
+  options = {
+
+    nix.optimise = {
+
+      automatic = mkOption {
+        default = false;
+        type = types.bool;
+        description = "Automatically run the nix store optimiser at a specific time.";
+      };
+
+      dates = mkOption {
+        default = "03:45";
+        type = types.str;
+        description = ''
+          Specification (in the format described by
+          <citerefentry><refentrytitle>systemd.time</refentrytitle>
+          <manvolnum>5</manvolnum></citerefentry>) of the time at
+          which the optimiser will run.
+        '';
+      };
+    };
+  };
+
+
+  ###### implementation
+
+  config = {
+
+    systemd.services.nix-optimise =
+      { description = "Nix Store Optimiser";
+        script = "exec ${config.nix.package.out}/bin/nix-store --optimise";
```

Comment:
removing `out` didn't work for me, I ended up with the `-dev` store path.  Are the "recent changes" you mention more recent than this commit's parent (https://github.com/NixOS/nixpkgs/commit/6b20d5b75efb3e4c95f7dabbfe0e607c1205f9be)?  If so maybe I need to wait for `nixos-unstable` to move forward a bit ...